purchases/src/main/kotlin/com/revenuecat/purchases/utils/OfferingImagePreDownloader.kt

```diff
@@ -27,14 +47,140 @@ internal class OfferingImagePreDownloader(
+
+    private fun downloadV1Images(offering: Offering) {
         offering.paywall?.let { paywallData ->
             val imageUris = paywallData.config.images.all.map {
                 Uri.parse(paywallData.assetBaseURL.toString()).buildUpon().path(it).build()
             }
             imageUris.forEach {
-                debugLog("Pre-downloading paywall image: $it")
+                debugLog("Pre-downloading Paywall V1 paywall image: $it")
+                coilImageDownloader.downloadImage(it)
+            }
+        }
+    }
+
+    private fun downloadV2Images(offering: Offering) {
+        offering.paywallComponents?.let { paywallComponents ->
+            val imageUrls = findImageUrisToDownload(paywallComponents)
+            imageUrls.forEach {
+                debugLog("Pre-downloading Paywall V2 paywall image: $it")
                 coilImageDownloader.downloadImage(it)
             }
         }
     }
+
+    private fun findImageUrisToDownload(paywallComponents: Offering.PaywallComponents): Set<Uri> {
+        val paywallComponentsConfig = paywallComponents.data.componentsConfig.base
+
+        return paywallComponentsConfig.stack.findImageUrisToDownload() +
+            (paywallComponentsConfig.stickyFooter?.stack?.findImageUrisToDownload() ?: emptyList()) +
+            paywallComponentsConfig.background.findImageUrisToDownload()
+    }
+
+    private fun StackComponent.findImageUrisToDownload(): Set<Uri> {
+        return filter {
+            it is StackComponent ||
+                it is IconComponent ||
+                it is CarouselComponent ||
+                it is TabsComponent ||
+                it is ImageComponent
+        }.flatMap { component ->
+            when (component) {
+                is StackComponent -> {
+                    component.background.findImageUrisToDownload() + component.overrides.flatMap { override ->
+                        override.properties.background.findImageUrisToDownload()
+                    }.toSet()
+                }
+                is IconComponent -> {
+                    setOf(Uri.parse(component.baseUrl).buildUpon().path(component.formats.webp).build())
+                }
+                is CarouselComponent -> {
+                    component.background.findImageUrisToDownload() + component.overrides.flatMap { override ->
+                        override.properties.background.findImageUrisToDownload()
+                    }.toSet()
+                }
+                is TabsComponent -> {
+                    component.background.findImageUrisToDownload() + component.overrides.flatMap { override ->
+                        override.properties.background.findImageUrisToDownload()
+                    }.toSet()
+                }
+                is ImageComponent -> {
+                    component.source.findImageUrisToDownload() + component.overrides.flatMap { override ->
+                        override.properties.source?.findImageUrisToDownload() ?: emptySet()
```

Comment:
Here `?: emptySet()` and `orEmpty()` are not very different (no extra parenthesis needed for instance), so feel free to ignore. 
```suggestion
                        override.properties.source?.findImageUrisToDownload().orEmpty()
```
(Also needs an import probably.)